### PR TITLE
Add grpc proxying to enable bring-your-own-proto

### PR DESCRIPTION
**Pull request by yaron2**
Closes https://github.com/dapr/dapr/issues/3231.
Closes https://github.com/dapr/dapr/issues/3096.
Closes https://github.com/dapr/dapr/issues/3065.

This PR enables users to work with their existing gRPC protos for service invocation, while still utilizing Dapr's service invocation features like mTLS, API auth, telemetry, access lists etc.

#### High level changes made in this PR

1. Access lists functionality *moved* from `pkg/configuration` to `pkg/acl`. This is for the better and ACLs now live in a dedicated and reusable package.
2. Adding Stream trace middleware for gRPC under `grpc_diagnostics.go` to support users with gRPC streams.
3. Added the use of a gRPC transparent handler that listens to requests for services that Dapr doesn't know about (user services). This is done through the user of [this](https://github.com/trusch/grpc-proxy) package which I've reviewed. Once the review process is done, I'll make another commit to migrate the relevant code to Dapr with appropriate attributions. The reason is to remove the dependency ownership.
4. Marked this feature as experimental, meaning it's opt-in until it graduates to a stable level.
5. Added unit tests

#### Two phase review

In order to keep the list of files changed small for the actual code change, I'm waiting with adding the e2e tests. Once the initial review is done, the following will be added:

1. E2E test
2. Code for the transparent proxy handler

### CODE REVIEW

**Comment by daixiang0** [comment]
Could you share more about why choose this grpc proxy and what changes have made?

File: pkg/diagnostics/grpc_tracing.go [hunk 1]
```diff
@@ -81,6 +69,72 @@ func GRPCTraceUnaryServerInterceptor(appID string, spec config.TracingSpec) grpc
...
 }
 
+// GRPCTraceStreamServerInterceptor sets the trace context or starts the trace client span based on request.
+func GRPCTraceStreamServerInterceptor(appID string, spec config.TracingSpec) grpc.StreamServerInterceptor {
+	return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		var span *trace.Span
+		spanName := info.FullMethod
+
+		ctx := ss.Context()
+		md, _ := metadata.FromIncomingContext(ctx)
```
Comment (daixiang0) [inline pkg/diagnostics/grpc_tracing.go:79]: add check if ok here.

File: pkg/diagnostics/grpc_tracing.go [hunk 2]
```diff
@@ -81,6 +69,72 @@ func GRPCTraceUnaryServerInterceptor(appID string, spec config.TracingSpec) grpc
...
+
+		vals := md.Get(GRPCProxyAppIDKey)
+		if len(vals) == 0 {
+			return errors.Errorf("cannot proxy request: missing %s metadata", GRPCProxyAppIDKey)
+		}
+
+		targetID := vals[0]
```
Comment (daixiang0) [inline pkg/diagnostics/grpc_tracing.go:86]: always one or only pick the first one?

File: pkg/grpc/server.go [hunk 3]
```diff
@@ -160,6 +166,10 @@ func (s *server) getMiddlewareOptions() []grpc_go.ServerOption {
 	if diag_utils.IsTracingEnabled(s.tracingSpec.SamplingRate) {
 		s.logger.Info("enabled gRPC tracing middleware")
 		intr = append(intr, diag.GRPCTraceUnaryServerInterceptor(s.config.AppID, s.tracingSpec))
+
+		if s.proxy != nil {
+			intrStream = append(intrStream, diag.GRPCTraceStreamServerInterceptor(s.config.AppID, s.tracingSpec))
```
Comment (artursouza) [inline pkg/grpc/server.go:171]: OK, so you are using the middleware code to intercept the gRPC calls and decide to proxy to the app. Nice.

File: pkg/runtime/runtime.go [hunk 4]
```diff
@@ -510,7 +516,18 @@ func (a *DaprRuntime) initDirectMessaging(resolver nr.Resolver) {
...
 		resolver,
 		a.globalConfig.Spec.TracingSpec,
-		a.runtimeConfig.MaxRequestBodySize)
+		a.runtimeConfig.MaxRequestBodySize,
+		a.proxy)
+}
+
+func (a *DaprRuntime) initProxy() {
+	//TODO: remove feature check once stable
+	if config.IsFeatureEnabled(a.globalConfig.Spec.Features, messaging.GRPCFeatureName) {
```
Comment (artursouza) [inline pkg/runtime/runtime.go:525]: Thanks.

**Review by artursouza — COMMENTED** [review]
LGTM. Please, address the comments above.

File: pkg/diagnostics/grpc_tracing.go (hunk 2 above)
Comment (yaron2) [reply pkg/diagnostics/grpc_tracing.go:86]: There's only one.

**Comment by yaron2** [comment]
> Could you share more about why choose this grpc proxy and what changes have made?

I chose this library as it's complete functionality wise, very well tested, and has no dependencies other than Go stdlib and grpc libraries.

**Comment by yaron2** [comment]
> LGTM. Please, address the comments above.

Ok, I'll now move into the second phase which is E2E and the porting of the proxy library.

**Comment by yaron2** [comment]
Proxy porting is done.

**Review by artursouza — APPROVED** [review]

**Pull request merged**